Version 1.0: North Atlantic Update - Tides, Curved Boundaries, Config, and More!😀🤯🥳 (#193)
COSIMA/regional-mom6: docs/angle_calculation.md

# Rotation and angle calculation in regional-mom6 using MOM6 Angle Calculation

Here we explain the implementation of MOM6 angle calculation in regional-mom6, which is the process by which regional-mom6 calculates the angle of curved horizontal grids (``hgrids``).

> **Issue:** On a curved hgrid, we have to rotate the boundary conditions according to the angle the grid is rotated from lat-lon coordinates (true north vs model north). MOM6 calculates the angle internally. ``rotation.py`` copies that MOM6 calculation into python so we can rotate it based on that calculation. The issue is that usually users provide hgrids in MOM6 with an angle field (``angle_dx``) and we were using that field. We don't necessarily trust that the user calculation doesn't have slight differences, so we implemented a way to use MOM6s angle calculation instead. In short, MOM6 doesn't actually use the user-provided ``angle_dx`` field in input hgrids, but internally calculates the angle. 

Reviewer: ```suggestion
**Issue:** On a curved ``hgrid``, we have to rotate the boundary conditions according to the local angle that the grid is rotated relative to latitude-longitude coordinates. MOM6 calculates the angle internally. ``regional-mom6`` replicates the MOM6 calculation. The issue is that usually users provide ``hgrid``s in MOM6 with an angle field (``angle_dx``). But there is no way to ensure that the user-provided ``angle_dx`` field was computed replicating the exact way MOM6 internally calculates the local rotation angle.
```

> **Solution:** To accomodate this fact, when we rotate our boundary conditions, we implemented MOM6 angle calculation in a file called "rotation.py", and adjusted functions where we regrid the boundary conditions.

Reviewer: ```suggestion
**Solution:** The exact angle calculation used internally by MOM6 is replicated in Python within "rotation.py" along-side with adjusted functions to help rotate and regrid the boundary conditions.
```

Reviewer 2: Idk how to add suggestions on @navidcy's suggestions, but I think you want either along with or alongside, not alongside with

Reviewer: changed it

## MOM6 process of angle calculation (T-point only)

> 1. Calculate pi/4rads / 180 degrees  = Gives a 1/4 conversion of degrees to radians. I.E. multiplying an angle in degrees by this gives the conversion to radians at 1/4 the value. 

Reviewer: I don't understand what step 1 tries to say. This doesn't read like a sentence? I think my main problem is the syntax, I am confused about what the sentence is trying to say..

4. We then find the lon_scale, which is the "trigonometric scaling factor converting changes in longitude to equivalent distances in latitudes". Whatever that actually means is we add the latitude of all four of these points from part 3 and basically average it and convert to radians. We then take the cosine of it. As I understand it, it's a conversion of longitude to equivalent latitude distance. 
5. Then we calculate the angle. This is a simple arctan2 so y/x. 
    1. The "y" component is the addition of the difference between the diagonals in longitude (adjusted by modulo_around_point in step 3) multiplied by the lon_scale, which is our conversion to latitude.
    2. The "x" component is the same addition of differences in latitude.
    3. Thus, given the same units, we can call arctan to get the angle in degrees

## Problem
MOM6 only calculates the angle at t-points. For boundary rotation, we need the angle at the boundary, which is q/u/v points. Because we need the points to the left, right, top, and bottom of the point, this method won't work for the boundary.

> 

Reviewer 2: This documentation is very detailed and I like that it's all explained properly! However maybe we should put in a tl;dr here so that casual users know they don't need to do anything special? 

```suggestion
## Default behaviour
The package defaults XYZ. The remainder of this page details the other options
```
Or something. @manishvenu I noticed though that the default behaviour is set to the given angle - since you've figured out that this is different than what MOM6 uses internally, should we use Fred's method by default? This addresses @navidcy 's concern about failsafing the code!

Author: Done!

Reviewer: hey both, I tried to understand this but I still have trouble it seems

@ashjbarnes, you are claiming that the documentation is clear. But currently, the documentation (https://regional-mom6--193.org.readthedocs.build/en/193/angle_calculation.html#boundary-rotation-algorithm) includes phrases like "whatever that actually means". This shows that we don't understand what's happening there, right? If we don't understand then why are we trying to document it?

Reviewer: While I was trying to understand the algorithm by reading the MOM6 source code so that I can document it nicely I found a bug in the implementation. And that bug wasn't captured by the tests.

The fact that we don't understand what the algorithm does and the fact that the tests were not actually capturing potential bugs made me a bit worried.

Author: > While I was trying to understand the algorithm by reading the MOM6 source code so that I can document it, I found a bug in the implementation. And that bug wasn't captured by the tests.

The bug was to the 1e-5 on the test -> it couldn't have been caught because the check we have is different enough angle calculation method that is larger than this difference. The test did catch the bug when we tried to fix it the first time (and we were wrong), so I think it'll work for major changes.

> The fact that we don't understand what the algorithm does and the fact that the tests were not actually capturing potential bugs made me a bit worried.

Feel free to take out all the uncertain parts of the docs if that helps.

Reviewer: Thanks. I will rewrite the doc. I need to understand the algorithm first to be able to explain it.

> ## Convert this method to boundary angles - 3 Options

Reviewer 2: There are only two options?

Reviewer: Yes, true.  Perhaps here we count the "do no rotation" as one of the three?

1. **GIVEN_ANGLE**: Don't calculate the angle and use the user-provided field in the hgrid called "angle_dx"
2. **EXPAND_GRID**: Calculate another boundary row/column points around the hgrid using simple difference techniques. Use the new points to calculate the angle at the boundaries. This works because we can now access the four points needed to calculate the angle, where previously at boundaries we would be missing at least two. 

## Code Description

Most calculation code is implemented in the rotation.py script, and the functional uses are in regrid_velocity_tracers and regrid_tides functions in the segment class of regional-mom6.

### Calculation Code (rotation.py)
1. **Rotational Method Definition**:  Rotational Methods are defined in the enum class "Rotational Method" in rotation.py.
2. **MOM6 Angle Calculation**: The method is implemented in "mom6_angle_calculation_method" in rotation.py and the direct t-point angle calculation is "initialize_grid_rotation_angle". 

> 3. **Fred's Pseudo Grid Expansion**: The method to add the additional boundary row/columns is referenced in "pseudo_hgrid" functions in rotation.py

Reviewer 2: ```suggestion
3. **Fred Castruccio's Pseudo Grid Expansion**: This method adds the additional boundary row/columns and is implemented in the `create_expanded_hgrid` function in `rotation.py`
```
This isn't right:  `pseudo_hgrid` is a variable not a function in the code. Perhaps you meant `create_expanded_hgrid`?

Reviewer: (I left this unresolved for the question to @manishvenu by @ashjbarnes.)

Author: Looks like this is resolved! No longer in the docs

### Implementation Code (regional_mom6.py)
Both regridding functions (regrid_velocity_tracers, regrid_tides) accept a parameter called "rotational_method" which takes the Enum class defining the rotational method.

We then define each method with a bunch of if statements. Here are the processes:

1. Given angle is the default method of accepting the hgrid's angle_dx

> 2. The EXPAND_GRID method is the least code, and we simply swap out the hgrid angle with the generated one we calculate right where we do the rotation.

Reviewer 2: The least code?
